### DISCUSSION TITLE
Multi-Site Permissions

### DISCUSSION THREAD

**Discussion by devilwacause** [area17/twill#2220] [Q&A]
## Description

Want to create a multi-site headless CMS with Twill.

We need to fine tune permissions like so.
     UserA -> can view siteB, edit/publish SiteA, edit/publish SiteB
     UserB -> can edit/publish SiteB, view SiteC, edit/publish SiteA
     UserC -> can edit SiteC, edit SiteA,  no access to SiteB
     etc.

How can we manage this with twill?

**Comment by ifox** [comment]
Hi @devilwacause this is possible in Twill 3 permissions system in combination with our multi-site capabilities: https://twill.io/docs/getting-started/configuration.html#content-multiple-subdomains-cms-routing. Groups give you the ability to allow a group of users to have access to 1 or more subdomains/sites. View/edit/publish permissions can be set at the role, group or user level, both globally per content type, or even per record if you need that granularity.

**Comment by devilwacause** [comment]
@ifox - Awesome.  
We were using twill v2 as we are searching for a compatible CMS to meet our needs.
I'll look into v3 this week.

Also - I've said before on a separate issue - you guys are amazingly quick to provide feedback and support here - which I want expressed - the community appreciates.

Now - just work on that documentation :p. ha

**Comment by devilwacause** [comment]
Set up multi-tenant.  

php artisan r:l

Shows the routes exist.
admin.{subdomain}.sdomain.domain.com/admin --  get 404 on attempt to reach it.

But what I need is a singular dashboard instead of sub-dashboards to manage sites. 

So 1 Dashboard - all models/resources associated with a site(s) (or globally associated)
Users limited to what they can see & do based on site assignment.

**Comment by devilwacause** [comment]
Disabling multi tenant - the routes work.

**Comment by devilwacause** [comment]
Still have multiple dashboards - but did figure out routing to meet our needs.

Incase somebody else is looking for this.

I needed so have routing as {subdomain}.maindomain.com  NOT admin.subdomain.maindomain.com if we were going to have multiple dashboards.  If you have the same needs - see below.

Replace this line in the following files.
FROM
`'domain' => config('twill.admin_app_subdomain', 'admin) . '.{subdomain}.' .....`
TO
`'domain' => '{subdomain}.' ....`

twill\src\RouteServiceProvider.php
twill\src\TwillRoutes.php


To make /admin (or whatever your chosen /base is) redirect to the correct login page.

in twill\src\Http\Middleware\Authenticate.php

in handle()
```
$host = $request->headers;
$host = $host->get('host');
$subdomain = explode('.', $host)[0];
return $request->expectsJson()
     ? abort(403, 'Your account is not verified.')
     : Redirect::guest(URL::route('twill.login.form', $subdomain));
```


in redirectTo
```
$host = $request->headers;
 $host = $host->get('host');
 $subdomain = explode('.', $host)[0];
 return route('twill.login.form', $subdomain);
```
